parser_library/src/workspaces/macro_cache.cpp

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2021 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+
+#include "macro_cache.h"
+
+#include <array>
+
+#include "file_manager.h"
+
+namespace hlasm_plugin::parser_library::workspaces {
+
+void macro_cache::collect_diags() const
+{
+    // No collectible children
+}
+
+macro_cache::macro_cache(const file_manager& file_mngr, file& macro_file)
+    : file_mngr_(&file_mngr)
+    , macro_file_(&macro_file)
+{}
+
+std::vector<cached_opsyn_mnemo> macro_cache_key::get_opsyn_state(context::hlasm_context& ctx)
+{
+    auto& wn = ctx.ids().well_known;
+    // List of instructions that are resolved during macro definition - therefore are affected by OPSYN
+    std::array<context::id_index, 19> cached_instr { wn.COPY,
+        wn.ASPACE,
+        wn.GBLA,
+        wn.GBLB,
+        wn.GBLC,
+        wn.LCLA,
+        wn.LCLB,
+        wn.LCLC,
+        wn.SETA,
+        wn.SETB,
+        wn.SETC,
+        ctx.ids().add("AIF"),
+        ctx.ids().add("MEND"),
+        ctx.ids().add("MACRO"),
+        ctx.ids().add("MEXIT"),
+        ctx.ids().add("AIF"),
+        ctx.ids().add("AREAD"),
+        ctx.ids().add("ACTR"),
+        ctx.ids().add("AGO") };
+
+    std::vector<cached_opsyn_mnemo> result;
+
+    for (const auto& [from, opcode] : ctx.opcode_mnemo_storage())
+    {
+        // If there is an opsyn, that aliases an instruction to be CA instruction, add it to result
+        if (std::find(cached_instr.begin(), cached_instr.end(), opcode.opcode) != cached_instr.end())
+            result.push_back(
+                { from, opcode.opcode, std::holds_alternative<context::macro_def_ptr>(opcode.opcode_detail) });
+
+        if (std::find(cached_instr.begin(), cached_instr.end(), from) != cached_instr.end())
+            result.push_back(
+                { from, opcode.opcode, std::holds_alternative<context::macro_def_ptr>(opcode.opcode_detail) });
+    }
+
+    // Also macros with the same name as CA instructions may alias the instructions
+    for (const auto& [id, macro] : ctx.macros())
```

Comment:
Are these inline defined macro only?

Reply:
No, the structure contains all macro definitions that were parsed up until the particular moment if parsing.